File: C#/p007-Categorical-Cross-Entropy-Loss.cs
/*
 * Calculating the loss with Categorical Cross Entropy
 * Based on nnfs.io by Harrison Kinsley & Daniel Kukiela
 * Associated with YT NNFS tutorial: https://www.youtube.com/watch?v=dEXPMQXoiLc
 * Contributed by T. Marin
 * 
 * Uses MathNet Numerics v4.15 by C.Ruegg, M. Cuda and J. Van Gael for vectors and matrix functions. 
 * To install in MSVS select Menu Project/Mangage NuGet packages...
 * in Browse tab search for MathNet.Numerics and install core package
*/

using System;
using MathNet.Numerics.LinearAlgebra;

namespace NNFS_p007
{
    class p007
    {
        static void Main(string[] args)
        {
            //creats shortcuts to Matrix and Vector builders: 
            var M = Matrix<double>.Build;
            var V = Vector<double>.Build;

            var softmax_output = M.DenseOfArray(new double[,] { { 0.7, 0.2, 0.1 } });
            var target_output = V.DenseOfArray(new double[] { 1, 0, 0 });

            var loss = - softmax_output.PointwiseLog() * target_output;

            Console.WriteLine($"{loss.ToString()}");

            Console.WriteLine($"{-Math.Log(0.7)}");

            Console.WriteLine($"{-Math.Log(0.5)}");

            Console.ReadLine();
        }
    }
}

File: Javascript/p007-Categorical-Cross-Entropy-Loss.js
/*
const b = 5.2;

console.log(Math.log(b));
console.log(Math.E ** 1.6486586255873816);
*/

const softmax_output = [0.7, 0.1, 0.2];
const target_output = [1, 0, 0];

const loss = -(Math.log(softmax_output[0])*target_output[0] +
			   Math.log(softmax_output[1])*target_output[1] +
			   Math.log(softmax_output[2])*target_output[2]);

console.log(loss);
// loss simplified
console.log(-Math.log(softmax_output[0]));

console.log(-Math.log(0.7));
console.log(-Math.log(0.5));
